torchdata/dataloader2/utils/worker.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import random
+
+from dataclasses import dataclass
+from typing import Callable, Optional
+
+import torch
+
+from torchdata.dataloader2.graph import DataPipe
+from torchdata.dataloader2.utils import generate_random_int
+from torchdata.datapipes.iter import IterDataPipe
+from torchdata.datapipes.map import MapDataPipe
+
+try:
+    import numpy
+
+    HAS_NUMPY = True
+except ModuleNotFoundError:
+    HAS_NUMPY = False
+
+
+@dataclass(frozen=True)
+class DistInfo:
+    r"""
+    Message class for keeping track of distributed information.
+
+    Args:
+        world_size (int): Total number of distributed nodes
+        rank (int): Distributed rank for the current distributed node
+    """
+    world_size: int = 1
+    rank: int = 0
+
+
+@dataclass(frozen=True)
+class WorkerInfo:
+    r"""
+    Message class for keeping track of worker information.
+
+    Args:
+        num_workers (int): Total number of worker processes
+        worker_id (int): Worker ID for the current worker process
+    """
+    num_workers: int
+    worker_id: int
+
+
+@dataclass(frozen=True)
+class _ExtraInfo:
+    r"""
+    Message class for extra arguments.
+    """
+    shared_seed: int
+
+
+def process_init_fn(
```

Comment:
We have to keep `process_init_fn` separately because `apply_sharding` can be called once per sharding-level cc: @NivekT